intel_extension_for_transformers/llm/runtime/graph/models/whisper/whisper.cpp

```diff
@@ -0,0 +1,4607 @@
+        const auto logprob = logprobs[i];
+        printf("%s : prob=%9.5f logit=%9.5f logprob=%9.5f\n", token.c_str(), prob, logit, logprob);
+    }
+
+    // "And", "and", " And", " and"
+    printf("logits[\"and\"]  = %f\n", logits[vocab.token_to_id.at("and")]);
+    printf("logits[\"And\"]  = %f\n", logits[vocab.token_to_id.at("And")]);
+    printf("logits[\" and\"] = %f\n", logits[vocab.token_to_id.at(" and")]);
+    printf("logits[\" And\"] = %f\n", logits[vocab.token_to_id.at(" And")]);
+    printf("logits[\" so\"]  = %f\n", logits[vocab.token_to_id.at(" so")]);
+
+    printf("logprobs[\"and\"]  = %f\n", logprobs[vocab.token_to_id.at("and")]);
+    printf("logprobs[\"And\"]  = %f\n", logprobs[vocab.token_to_id.at("And")]);
+    printf("logprobs[\" and\"] = %f\n", logprobs[vocab.token_to_id.at(" and")]);
+    printf("logprobs[\" And\"] = %f\n", logprobs[vocab.token_to_id.at(" And")]);
+    printf("logprobs[\" so\"]  = %f\n", logprobs[vocab.token_to_id.at(" so")]);
+
+    printf("probs[\"and\"]  = %f\n", probs[vocab.token_to_id.at("and")]);
+    printf("probs[\"And\"]  = %f\n", probs[vocab.token_to_id.at("And")]);
+    printf("probs[\" and\"] = %f\n", probs[vocab.token_to_id.at(" and")]);
+    printf("probs[\" And\"] = %f\n", probs[vocab.token_to_id.at(" And")]);
+    printf("probs[\" so\"]  = %f\n", probs[vocab.token_to_id.at(" so")]);
+#endif
+}
+
+static whisper_token_data whisper_sample_token(whisper_context& ctx, whisper_state& state,
+                                               const whisper_decoder& decoder, bool best) {
+  whisper_token_data result = {
+      0, 0, 0.0f, 0.0f, 0.0f, 0.0f, -1, -1, 0.0f,
+  };
+
+  const auto& vocab = ctx.vocab;
+
+  const auto& probs = decoder.probs;
+  const auto& logprobs = decoder.logprobs;
+
+  const int n_logits = vocab.n_vocab;
+
+  {
+    double sum_ts = 0.0;
+    double max_ts = 0.0;
+
+    for (int i = vocab.token_beg; i < n_logits; i++) {
+      if (probs[i] == -INFINITY) {
+        continue;
+      }
+
+      sum_ts += probs[i];
+      if (max_ts < probs[i]) {
+        max_ts = probs[i];
+        result.tid = i;
+      }
+    }
+
+    result.pt = max_ts / (sum_ts + 1e-10);
+    result.ptsum = sum_ts;
+  }
+
+  if (best) {
+    for (int i = 0; i < n_logits; ++i) {
+      if (result.p < probs[i]) {
+        result.id = i;
+        result.p = probs[i];
+        result.plog = logprobs[i];
+      }
+    }
+  } else {
+    std::discrete_distribution<> dist(probs.begin(), probs.end());
+
+    result.id = dist(state.rng);
+    result.p = probs[result.id];
+    result.plog = logprobs[result.id];
+  }
+
+  if (result.id >= vocab.token_beg) {
+    result.tid = result.id;
+    result.pt = result.p;
+  }
+
+  state.n_sample++;
+
+  return result;
+}
+
+static std::vector<whisper_token_data> whisper_sample_token_topk(whisper_context& ctx, whisper_state& state,
+                                                                 const whisper_decoder& decoder, int k) {
```

Comment:
seems all post processing are within whisper itself, can we reuse these effort under ```model_utils```?

Reply:
@zhentaoyu correct me

Reply:
I could try latter.

Reply:
I'm not familiar with `whisper`, just know it's a [speech-to-text](https://openai.com/research/whisper) model. So why does it need `top_k` sampling rather than `top-1`?
![image](https://github.com/intel/intel-extension-for-transformers/assets/84065714/05f26d72-5020-445f-9073-cc2f43433bce)